src/main/java/com/direwolf20/buildinggadgets/common/packets/UpdateTemplatePacket.java

```diff
@@ -0,0 +1,211 @@
+package com.direwolf20.buildinggadgets.common.packets;
+
+import com.direwolf20.buildinggadgets.BuildingGadgets;
+import com.direwolf20.buildinggadgets.common.capbility.TemplateProviderCapability;
+import com.direwolf20.buildinggadgets.common.schema.template.Template;
+import com.direwolf20.buildinggadgets.common.schema.template.TemplateIO;
+import com.google.common.base.MoreObjects;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import net.minecraft.entity.Entity;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.nbt.CompressedStreamTools;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.world.World;
+import net.minecraftforge.fml.network.NetworkEvent.Context;
+import net.minecraftforge.fml.network.PacketDistributor.PacketTarget;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+
+public final class UpdateTemplatePacket extends RequestTemplatePacket {
+    private static final int SPLIT_SIZE = 30000;//split every 30000 bytes - this is below the client limit
+    private static final AtomicInteger curSession = new AtomicInteger();
+    //null UUID indicates server-origin assembly
+    private static final LoadingCache<Optional<UUID>, PacketAssembly> assemblyByPlayerId = CacheBuilder.newBuilder()
+            .concurrencyLevel(2)
+            .expireAfterAccess(5, TimeUnit.MINUTES)
+            .build(new CacheLoader<Optional<UUID>, PacketAssembly>() {
+                @Override
+                public PacketAssembly load(@Nonnull Optional<UUID> key) {
+                    return new PacketAssembly();
+                }
+            });
+
+    public static void send(UUID id, @Nullable Template template, PacketTarget target) {
+        Optional.ofNullable(template)
```

Comment:
I'm not a fan of this at all. Please don't inject functional code like this when it doesn't solve anything. I personally find this kind of code more messy.

Reply:
Well it avoids a bunch of annoying and confusing if's. This code makes the actual algorithm clearer (even though not having ifPresentOrElse is really annoying.

Reply:
It also solves having to duplicate the null Template Update-packet with a more elegant solution then having another Method.

Reply:
can we move the .map into a method at the very least so it at least has a name to it. I just love blocks of uncommented code.

Reply:
Going to go for 2 Methods then.